### PR TITLE
AST generation methods refactor

### PR DESCRIPTION
**Pull request by GeoffreyBooth**
Sorry this has taken so long, I needed to find the time to really dig into this. So, code like this:

```coffee
  withAstType: (ast, o) ->
    return ast unless ast
    return ast if Array.isArray ast
    return ast if ast.type
    return ast unless @emptyAst or do ->
      return yes for key in Object.keys(ast) when key not in ['comments', astLocationFields...]

    merge ast,
      type: @astType?(o) ? @astType
```

Was bothering me. On a basic level, we shouldn’t be passing around the AST that we’re generating, for other functions to iterate over it and mutate it. That would be hell to debug whenever the final AST output isn’t what we’re expecting. But on the broader level, this doesn’t follow object-oriented principles. We shouldn’t be passing in an output object into each class for the class to decorate the nodes of the output with extra properties. This is a helper function attached to a class as a method.

All the base AST methods felt like this to me, but rather than giving you annoying feedback like “make this more object-oriented” or “follow the pattern of the node classes” I thought I should take a crack at refactoring it myself. Which I did, and the tests still pass. So here’s the basic idea:

```coffee
  ast: ->
    # Every abstract syntax tree node object has four categories of properties:
    # - type, stored in the `type` field and a string like `NumberLiteral`.
    # - location data, stored in the `loc`, `start`, `end` and `range` fields.
    # - properties specific to this node, like `parsedValue`.
    # - properties that are themselves child nodes, like `body`.
    # These fields are all intermixed in the Babel spec; `type` and `start` and
    # `parsedValue` are all top level fields in the AST node object. We have
    # separate methods for returning each category, that we merge together here.
    Object.assign {}, @astProperties(), {type: @astType()}, @astLocationData()
```

I renamed `toAst` to `ast` so that all our AST-related methods start with `ast`. Still not happy with naming but that’s another story. The basic idea here is that there are three methods that return all the parts of an AST object: `astProperties`, `astType` and `astLocationData`. Of the three, `astType` returns a string and the others return objects—always. No complicated `if` trees checking if `astProperties` is a function or array or object, it should just always be a function that always returns an object, and that’s it.

And there’s no looping in the base class. Children get added by overriding `astProperties` for nodes that have child nodes, such as `Block`:

```coffee
  astProperties: ->
    expressions: @expressions.map (child) => child.ast()
```

And that’s it. Thoughts?

### CODE REVIEW

File: src/nodes.coffee [hunk 1]
```diff
@@ -269,98 +268,43 @@ exports.Base = class Base
...
-  # We try to follow the [Babel AST spec](https://github.com/babel/babel/blob/master/packages/babylon/ast/spec.md)
+  # We try to follow the [Babel AST spec](https://github.com/babel/babel/blob/master/packages/babel-parser/ast/spec.md)
   # as closely as possible, for improved interoperability with other tools.
-  toAst: (o, level) ->
-    o = extend {}, o
-    o.level = level if level
-    @withAstLocationData @withAstType @getAstContent(o), o
-
-  # By default, a node class's AST `type` is the class name
+  ast: ->
```
Comment (helixbass) [inline src/nodes.coffee:273]: This will need to accept `o` as its first argument (the `nodes.ast options` call in `coffeescript.coffee` passes it in initially)

File: src/nodes.coffee [hunk 2]
```diff
@@ -269,98 +268,43 @@ exports.Base = class Base
...
+    # Every abstract syntax tree node object has four categories of properties:
+    # - type, stored in the `type` field and a string like `NumberLiteral`.
+    # - location data, stored in the `loc`, `start`, `end` and `range` fields.
+    # - properties specific to this node, like `parsedValue`.
+    # - properties that are themselves child nodes, like `body`.
+    # These fields are all intermixed in the Babel spec; `type` and `start` and
+    # `parsedValue` are all top level fields in the AST node object. We have
+    # separate methods for returning each category, that we merge together here.
+    Object.assign {}, @astProperties(), {type: @astType()}, @astLocationData()
```
Comment (helixbass) [inline src/nodes.coffee:282]: ...and then some of these will end up needing to pass/accept `o` as well for cases where it affects the AST, `@astProperties(o)` and probably `@astType(o)` at least

It's definitely fine (and perhaps instructive) to not add these things until a PR for a specific node class needs them

File: test/abstract_syntax_tree_location_data.coffee [hunk 3]
```diff
@@ -1,8 +1,29 @@
...
+    "Expected location start column #{reset}#{node.loc.start.column}#{red} to equal #{reset}#{expected.loc.start.column}#{red}"
+  eq node.loc.end.line, expected.loc.end.line, \
+    "Expected location end line #{reset}#{node.loc.end.line}#{red} to equal #{reset}#{expected.loc.end.line}#{red}"
+  eq node.loc.end.column, expected.loc.end.column, \
+    "Expected location end column #{reset}#{node.loc.end.column}#{red} to equal #{reset}#{expected.loc.end.column}#{red}"
+
 
 test "AST location data as expected for NumberLiteral node", ->
-  testExpression '42',
-    type: 'NumericLiteral'
```
Comment (helixbass) [inline test/abstract_syntax_tree_location_data.coffee]: The reason I had the `type`s here was to verify that we're actually examining the node we think we are. The `node = ast.expressions[0]` that is being used to say "we're examining the first thing in this AST" works for now, but won't hold up for long

Eg a real full top-level (file) AST for `42` will end up looking like
```
type: 'File'
program:
  type: 'Program'
  body: [
    type: 'ExpressionStatement'
    expression:
      type: 'NumericLiteral'
  ]
```

Perhaps an elegant mechanism for specify which node the location data belongs to would be to have `testAstLocationData()` accept eg a `path:` argument that describes the path to the desired node? And it could default to the path to the first expression in the `program.body`?

Eg to check the location data for the test of a top-level `if` statement, you could do something like `path: program[0].body.test` (or maybe the initial prefix can be omitted)

Regardless of that (and it's not really necessary yet), I don't see the harm in including a `type:` in these tests, I think all it does is add clarity to what is under test, no?

**Review by helixbass — APPROVED** [review]
@GeoffreyBooth ok I like some of the simplifications. My thought of how to proceed is:
- I'd prepared a couple "leaf" node class AST PRs, why don't I go ahead and rebase them onto this version of this branch and try out the new AST generation "API" a little bit that way, and I'll put those up for you to look at as well
- Then I've been merging these PR branches back into my `prettier` branch to keep them in sync, adjusting that branch to use this modified API will take some time but will be a good way to test out my biggest question which is "what will it look like for node classes that have more complex AST generation code"

> there are three methods that return all the parts of an AST object: `astProperties`, `astType` and `astLocationData`

I can definitely get behind merging `astChildren()` and `astProps()` into just `astProperties()`. And having `astLocationData()` as a method makes sense

> Of the three, `astType` returns a string and the others return objects—always. No complicated `if` trees checking if `astProperties` is a function or array or object, it should just always be a function that always returns an object, and that’s it.

This is fine with me. I personally lean more towards flexible interfaces that give you multiple ways of "declaring" what you want, but I can see the simplicity of having a known style for calling them

So basically I'm slightly tentatively ok with proceeding with this as-is. There'll be questions that come up as there's more complexity in some real use cases, but I can see how it's valuable (both as an exercise and a way for me to reexamine the choices I've made on my `prettier` branch) to deal with those as they come up

**Comment by GeoffreyBooth** [comment]
Glad you like it! Yeah, I was thinking that this might be hell to rebase against because of all the work you’ve done that goes beyond this, and I appreciate the effort you’ll put into updating the other branches. But I think this is worth spending the time to get right, especially if it turns into the biggest refactor of the CoffeeScript codebase in years, which is the direction it seems like we’re heading.

And I’m not 100% sure that this is _the_ architecture, not by a long shot. I’m sure this will get more complicated as we start adding more complex node types. But I think there’s a lot to be said for starting simple.

> This is fine with me. I personally lean more towards flexible interfaces that give you multiple ways of “declaring” what you want, but I can see the simplicity of having a known style for calling them

The “known style” approach is the current style of the node classes; just look at `YES = -> yes` near the top for an example, where there’s a shorthand for a function that returns `true` rather than the calling code being able to handle either a boolean value or a function. I think for a codebase like this one where a lot of hands will touch it over years, it’s better to be prescriptive rather than flexible, to maintain consistency.

As for passing `o`, it just didn’t seem to be necessary in this PR. I initially had it, but took it out when it didn’t seem to be getting used. I agree, I’ll be curious to see the first case where we need it, to dig in and determine why.

I see your point about checking the type in the location data tests. I restored that.

File: test/abstract_syntax_tree_location_data.coffee [hunk 4]
```diff
@@ -1,8 +1,29 @@
-testExpression = expressionAstMatchesObject
+# Astract Syntax Tree location data
+# ---------------------------------
+
+testAstLocationData = (code, expected) ->
+  ast = CoffeeScript.compile code, ast: yes
+  # Pull the node we’re testing out of the root `Block` node’s first child.
+  node = ast.expressions[0]
```
Comment (helixbass) [inline test/abstract_syntax_tree_location_data.coffee:7]: So this was the most problematic thing I ran into rebasing my `Value` AST branch onto this branch: I had written location data tests for eg `a.b` that include location data for nested nodes (not just the root node)

And now I notice that per my other comment that `node = ast.expressions[0]` is not very "solid", I don't think it's good to separately hard-code it here. What exactly was gained by removing those shared test helpers?

If it's about having the nicer error messages here, you could probably refactor so that it recursively looked for nodes with location data in the "expected" location data object and then applied the individual location data field tests (with custom error messages) to each node with location data that it found?

Let me know if you'd rather do that refactoring here or if I should come up with something in the `Value` AST PR

**Review by helixbass — COMMENTED** [review]
@GeoffreyBooth I rebased my `Value` AST branch onto this branch and opened https://github.com/jashkenas/coffeescript/pull/5107. Overall was quite smooth using the new AST methods, just see one comment about testing (multiple) nested location data

File: test/abstract_syntax_tree_location_data.coffee (hunk 4 above)
Comment (GeoffreyBooth) [reply test/abstract_syntax_tree_location_data.coffee:7]: My concern was about code organization, of keeping all the AST stuff close together rather than adding more functions to `helpers.coffee` where those functions are only used by the two AST files. We could move them into `test/support/abstract_syntax_tree_helpers.coffee`, I suppose, and update `Cakefile` to load them like it currently makes the generic helpers global. Another option would be to just have one big AST tests file, rather than a separate one for location data tests, and then all the helpers and tests are in the same file.

Comment (helixbass) [reply test/abstract_syntax_tree_location_data.coffee:7]: Ok I think I like option 1 (`test/support/abstract_syntax_tree_helpers.coffee`) at this point, as I anticipate both the main and location data AST test files getting quite large so seems nice in theory to keep them separate

Updated https://github.com/jashkenas/coffeescript/pull/5107 (including `test/support/abstract_syntax_tree_helpers.coffee`) to support the test cases there

Comment (GeoffreyBooth) [reply test/abstract_syntax_tree_location_data.coffee:7]: Maybe let’s do merge in the test helpers refactor (the most recent one) in the Value PR? It doesn’t really matter which one it goes in, but I think it might make more sense to do it there because that’s the PR that necessitates it.

Then I think we can merge this in? And I’ll take one last look at the PR on the CoffeeScript repo, but I assume that that’s probably ready to merge in too.

Comment (helixbass) [reply test/abstract_syntax_tree_location_data.coffee:7]: Yup that's where I did the refactor in the Value PR (on `value-ast` branch)

**Review by helixbass — COMMENTED** [review]
@GeoffreyBooth merging
